sdk/src/main/java/io/dapr/exceptions/DaprErrorDetails.java

```diff
@@ -0,0 +1,223 @@
+    }
+
+    Map<ErrorDetailType, Map<String, Object>> detailsList = new HashMap<>();
+    List<Any> grpcDetailsList = status.getDetailsList();
+    for (Any detail : grpcDetailsList) {
+      for (Map.Entry<Class<? extends Message>, ErrorDetailType>
+          supportedClazzAndType : SUPPORTED_ERROR_TYPES.entrySet()) {
+        Class<? extends Message> clazz = supportedClazzAndType.getKey();
+        ErrorDetailType errorDetailType = supportedClazzAndType.getValue();
+        if (detail.is(clazz)) {
+          detailsList.put(errorDetailType, parseProtoMessage(detail, clazz));
+        }
+      }
+    }
+    return Collections.unmodifiableMap(detailsList);
+  }
+
+  private static Map<ErrorDetailType, Map<String, Object>> parse(List<Map<String, Object>> entries) {
+    if ((entries == null) || entries.isEmpty()) {
+      return Collections.emptyMap();
+    }
+
+    Map<ErrorDetailType, Map<String, Object>> detailsList = new HashMap<>();
+    for (Map<String, Object> entry : entries) {
+      Object type = entry.getOrDefault("@type", "");
+      if (type == null) {
+        continue;
+      }
+
+      Class<? extends Message> clazz = ERROR_TYPES_FQN_REVERSE_LOOKUP.get(type.toString());
+      if (clazz == null) {
+        continue;
+      }
+
+      ErrorDetailType errorDetailType = SUPPORTED_ERROR_TYPES.get(clazz);
+      if (errorDetailType == null) {
+        continue;
+      }
+
+      detailsList.put(errorDetailType, entry);
+    }
+    return Collections.unmodifiableMap(detailsList);
+  }
+
+  private static <T extends com.google.protobuf.Message> Map<String, Object> parseProtoMessage(
+      Any detail, Class<T> clazz) {
+    try {
+      T message = detail.unpack(clazz);
+      return messageToMap(message);
+    } catch (InvalidProtocolBufferException e) {
+      return Collections.singletonMap(e.getClass().getSimpleName(), e.getMessage());
+    }
+  }
+
+  /**
+   * Converts a Protocol Buffer (proto) message to a Map.
+   *
+   * @param message The proto message to be converted.
+   * @return A Map representing the fields of the proto message.
+   */
+  private static Map<String, Object> messageToMap(Message message) {
+    Map<String, Object> result = new HashMap<>();
+    Field[] fields = message.getClass().getDeclaredFields();
+
+    result.put("@type", generateErrorTypeFqn(message.getClass()));
+
+    for (Field field : fields) {
+      if (field.isSynthetic() || Modifier.isStatic(field.getModifiers())) {
+        continue;
+      }
+
+      String normalizedFieldName = field.getName().replaceAll("_$", "");
+      try {
+        field.setAccessible(true);
+        Object value = field.get(message);
+        result.put(normalizedFieldName, value);
+      } catch (IllegalAccessException e) {
+        // no-op, just ignore this attribute.
+      }
+    }
+
+    return Collections.unmodifiableMap(result);
+  }
+
+  private static <T extends com.google.protobuf.Message> String generateErrorTypeFqn(Class<T> clazz) {
```

Comment:
Clarifying my change: just a small refactoring to allow this logic to be reused.